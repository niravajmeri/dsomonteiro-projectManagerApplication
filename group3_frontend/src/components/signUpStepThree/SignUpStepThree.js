import React, { Component } from "react";
import {
    Button,
    FormGroup,
    FormControl,
    Alert,
    ControlLabel
} from "react-bootstrap";

class SignUpStepThree extends Component {
    constructor(props) {
        super(props);
        this.state = {
            verificationCode: "",
            hideWrongCode: "hide-code"
        };
    }

    handleCodeSubmit = event => {
        event.preventDefault();
        const verificationCode = this.state.verificationCode;
        fetch(this.props.validationCodeUrl, {
            body: JSON.stringify({ codeToCheck: verificationCode }),
            headers: {
                "content-type": "application/json"
            },
            method: "POST"
        })
            .then(response => response.json())
            .then(jsonResponse => {
                this.props.setStepThree({
                    signupStep: 4
                });
            })
            .catch(error => {
                this.setState({
                    hideWrongCode: ""
                });
                console.error("Verification code is incorrect. Try again");
            });
    };

    validateCodeForm() {
        return this.state.verificationCode.length > 0;
    }

    handleChange = event => {
        this.setState({
            [event.target.id]: event.target.value
        });
    };

    stepThreeForm = () => (
        <div className="validationCodeForm">
            <form onSubmit={this.handleCodeSubmit}>
                <FormGroup controlId="verificationCode">
                    <ControlLabel>Validation code</ControlLabel>
                    <FormControl
                        autoFocus
                        type="text"
                        value={this.state.verificationCode}
                        onChange={this.handleChange}
                    />
                </FormGroup>
                <Alert bsStyle="danger" className={this.state.hideWrongCode}>
                    <strong>Invalid validation code!</strong> Try again.
                </Alert>
                <Button block disabled={!this.validateCodeForm} type="submit">
                    Send validation code
                </Button>
            </form>
        </div>
    );

    render() {
        return this.stepThreeForm();
    }
}

export default SignUpStepThree;
