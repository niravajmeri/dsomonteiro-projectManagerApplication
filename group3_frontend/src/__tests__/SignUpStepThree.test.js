import React from 'react';
import { shallow } from 'enzyme';
import SignUpStepThree from '../components/signUpStepThree/SignUpStepThree';

describe('SignUpStepThree', () => {
    const fetchResult = {
        json: jest.fn().mockReturnValue({ href: 'responseUrl' })
    };

    let mockedClass;
    let eventMock = { preventDefault: jest.fn() };

    afterEach(() => {
        global.fetch.mockClear();
    });

    beforeEach(() => {
        global.fetch = jest
            .fn()
            .mockImplementation(() => Promise.resolve(fetchResult));

        mockedClass = new SignUpStepThree();
    });

    it('should fetch data with a valid verification code', () => {
        mockedClass.props = {
            validationCodeUrl: 'codeUrl',
            setStepThree: jest.fn()
        };
        mockedClass.state.verificationCode = 'code';

        mockedClass.handleCodeSubmit(eventMock).then(() => {
            const expectedProps = {
                signUpStep: 4
            };
            expect(mockedClass.props.setStepThree).toHaveBeenCalledWith(
                expectedProps
            );

            expect(global.fetch.mock.calls.length).toEqual(1);
            expect(global.fetch.mock.calls[0][0]).toBe('codeUrl');
            expect(global.fetch.mock.calls[0][1]).toEqual({
                body: '{"codeToCheck":"code"}',
                headers: { 'content-type': 'application/json' },
                method: 'POST'
            });

            expect(eventMock.preventDefault).toBeCalled();
        });
    });

    it('should setState when verification code is invalid', () => {
        global.fetch = jest.fn().mockImplementation(() => Promise.reject());
        mockedClass.setState = jest.fn();

        mockedClass.props = {
            validationCodeUrl: 'codeUrl'
        };
        mockedClass.state.verificationCode = 'code';

        mockedClass.handleCodeSubmit(eventMock).then(() => {
            expect(mockedClass.setState).toBeCalledWith({
                hideWrongCode: ''
            });
        });
    });

    it('should render correctly', () => {
        const component = shallow(<SignUpStepThree />);
        expect(component).toMatchSnapshot();
    });

    it('should return true when the length of verificationCode is > 0', () => {
        mockedClass.state.verificationCode = '1234';
        expect(mockedClass.validateCodeForm()).toBeTruthy;
    });

    it('should return false when the length of verificationCode is < 0', () => {
        mockedClass.state.verificationCode = '';
        expect(mockedClass.validateCodeForm()).toBeFalsy;
    });

    it('handle change', () => {
        eventMock = {
            target: {
                id: 'id',
                value: 'value'
            }
        };
        mockedClass.setState = jest.fn();
        mockedClass.handleChange(eventMock);
        expect(mockedClass.setState).toHaveBeenLastCalledWith({ id: 'value' });
    });
});
