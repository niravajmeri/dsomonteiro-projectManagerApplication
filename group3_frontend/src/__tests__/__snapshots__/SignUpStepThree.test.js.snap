// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpStepThree should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignUpStepThree />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
>
        <FormGroup
                bsClass="form-group"
                controlId="verificationCode"
        >
                <ControlLabel
                        bsClass="control-label"
                        srOnly={false}
                >
                        Validation code
                </ControlLabel>
                <FormControl
                        autoFocus={true}
                        bsClass="form-control"
                        componentClass="input"
                        onChange={[Function]}
                        type="text"
                        value=""
                />
        </FormGroup>
        <Alert
                bsClass="alert"
                bsStyle="danger"
                className="hide-code"
                closeLabel="Close alert"
        >
                <strong>
                        Invalid validation code!
                </strong>
                 Try again.
        </Alert>
        <Button
                active={false}
                block={true}
                bsClass="btn"
                bsStyle="default"
                disabled={false}
                type="submit"
        >
                Send validation code
        </Button>
</form>,
      "className": "validationCodeForm",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <FormGroup
            bsClass="form-group"
            controlId="verificationCode"
>
            <ControlLabel
                        bsClass="control-label"
                        srOnly={false}
            >
                        Validation code
            </ControlLabel>
            <FormControl
                        autoFocus={true}
                        bsClass="form-control"
                        componentClass="input"
                        onChange={[Function]}
                        type="text"
                        value=""
            />
</FormGroup>,
          <Alert
            bsClass="alert"
            bsStyle="danger"
            className="hide-code"
            closeLabel="Close alert"
>
            <strong>
                        Invalid validation code!
            </strong>
             Try again.
</Alert>,
          <Button
            active={false}
            block={true}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            type="submit"
>
            Send validation code
</Button>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": Array [
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
>
                Validation code
</ControlLabel>,
              <FormControl
                autoFocus={true}
                bsClass="form-control"
                componentClass="input"
                onChange={[Function]}
                type="text"
                value=""
/>,
            ],
            "controlId": "verificationCode",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "control-label",
                "children": "Validation code",
                "srOnly": false,
              },
              "ref": null,
              "rendered": "Validation code",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "autoFocus": true,
                "bsClass": "form-control",
                "componentClass": "input",
                "onChange": [Function],
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "alert",
            "bsStyle": "danger",
            "children": Array [
              <strong>
                Invalid validation code!
</strong>,
              " Try again.",
            ],
            "className": "hide-code",
            "closeLabel": "Close alert",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Invalid validation code!",
              },
              "ref": null,
              "rendered": "Invalid validation code!",
              "type": "strong",
            },
            " Try again.",
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": false,
            "block": true,
            "bsClass": "btn",
            "bsStyle": "default",
            "children": "Send validation code",
            "disabled": false,
            "type": "submit",
          },
          "ref": null,
          "rendered": "Send validation code",
          "type": [Function],
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
>
          <FormGroup
                    bsClass="form-group"
                    controlId="verificationCode"
          >
                    <ControlLabel
                              bsClass="control-label"
                              srOnly={false}
                    >
                              Validation code
                    </ControlLabel>
                    <FormControl
                              autoFocus={true}
                              bsClass="form-control"
                              componentClass="input"
                              onChange={[Function]}
                              type="text"
                              value=""
                    />
          </FormGroup>
          <Alert
                    bsClass="alert"
                    bsStyle="danger"
                    className="hide-code"
                    closeLabel="Close alert"
          >
                    <strong>
                              Invalid validation code!
                    </strong>
                     Try again.
          </Alert>
          <Button
                    active={false}
                    block={true}
                    bsClass="btn"
                    bsStyle="default"
                    disabled={false}
                    type="submit"
          >
                    Send validation code
          </Button>
</form>,
        "className": "validationCodeForm",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <FormGroup
              bsClass="form-group"
              controlId="verificationCode"
>
              <ControlLabel
                            bsClass="control-label"
                            srOnly={false}
              >
                            Validation code
              </ControlLabel>
              <FormControl
                            autoFocus={true}
                            bsClass="form-control"
                            componentClass="input"
                            onChange={[Function]}
                            type="text"
                            value=""
              />
</FormGroup>,
            <Alert
              bsClass="alert"
              bsStyle="danger"
              className="hide-code"
              closeLabel="Close alert"
>
              <strong>
                            Invalid validation code!
              </strong>
               Try again.
</Alert>,
            <Button
              active={false}
              block={true}
              bsClass="btn"
              bsStyle="default"
              disabled={false}
              type="submit"
>
              Send validation code
</Button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": Array [
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
>
                  Validation code
</ControlLabel>,
                <FormControl
                  autoFocus={true}
                  bsClass="form-control"
                  componentClass="input"
                  onChange={[Function]}
                  type="text"
                  value=""
/>,
              ],
              "controlId": "verificationCode",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "control-label",
                  "children": "Validation code",
                  "srOnly": false,
                },
                "ref": null,
                "rendered": "Validation code",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "autoFocus": true,
                  "bsClass": "form-control",
                  "componentClass": "input",
                  "onChange": [Function],
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "alert",
              "bsStyle": "danger",
              "children": Array [
                <strong>
                  Invalid validation code!
</strong>,
                " Try again.",
              ],
              "className": "hide-code",
              "closeLabel": "Close alert",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Invalid validation code!",
                },
                "ref": null,
                "rendered": "Invalid validation code!",
                "type": "strong",
              },
              " Try again.",
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "active": false,
              "block": true,
              "bsClass": "btn",
              "bsStyle": "default",
              "children": "Send validation code",
              "disabled": false,
              "type": "submit",
            },
            "ref": null,
            "rendered": "Send validation code",
            "type": [Function],
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
